frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -2014,7 +2014,16 @@ static QualifiedType getProperFormalType(const ResolutionResultByPostorderID& r,
   return type;
 }
 
-static bool allowPromotionForSig(const TypedFnSignature* sig) {
+// if the signature is a function and has an AST, fn should be that AST.
+// This is used to do some other checking, like observing return intents.
+static bool allowPromotionForSig(const TypedFnSignature* sig,
+                                 const uast::Function* fn) {
+  // Functions that return 'param' or 'type' cannot be promoted
+  if (fn && (fn->returnIntent() == uast::Function::TYPE ||
```

Comment:
could just check `sig->untyped()->kind()` here so you don't need to get and pass in the actual `Function` ast to this function

Reply:
Oh, that's much better.

Reply:
Unfortunately, `kind()` is the function kind: `OPERATOR,` `ITER`, etc., and not the return intent. Looks like to get the return intent you need to either have a `ResolvedFunction` or the AST.

Reply:
Anna suggested including the return intent in the untyped signature, which I think is a good idea, but i will forego doing that in this PR since it is more invasive.